### PR TITLE
new-app: warn when source credentials are needed

### PR DESCRIPTION
**Pull request by csrwng**
When showing new-app results, displays a warning if a secret may be needed to access a git repository.

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees ptal

File: pkg/generate/app/cmd/describe.go [hunk 1]
```diff
@@ -200,6 +200,11 @@ func describeBuildPipelineWithImage(out io.Writer, ref app.ComponentReference, p
 				fmt.Fprintf(out, "      * Use 'start-build' to trigger a new build\n")
 			}
 		}
+
+		if pipeline.Build.Source.RequiresAuth {
+			fmt.Fprintf(out, "      * WARNING: this source repository may require credentials.\n"+
+				"                 Create a secret with your credentials and use 'set build-secret' to assign it to the build config.\n")
```
Comment (bparees) [inline pkg/generate/app/cmd/describe.go]: with your git credentials?

Comment (csrwng) [inline pkg/generate/app/cmd/describe.go]: will fix

**Comment by bparees** [comment]
lgtm otherwise.

File: pkg/generate/app/sourcelookup.go [hunk 2]
```diff
@@ -250,6 +254,50 @@ func (r *SourceRepository) LocalPath() (string, error) {
...
+			fmt.Sprintf("ProgramData=%s", os.Getenv("ProgramData"),
+			fmt.Sprintf("SystemRoot=%s", os.Getenv("SystemRoot"),
+		)
+	}
+	gitRepo := git.NewRepositoryWithEnv(env)
+	url, ok, err := r.RemoteURL()
+	if err != nil {
+		return err
+	}
+	if !ok {
```
Comment (smarterclayton) [inline pkg/generate/app/sourcelookup.go]: You should do this check before you create TempDirs

File: pkg/generate/app/sourcelookup.go [hunk 3]
```diff
@@ -250,6 +254,50 @@ func (r *SourceRepository) LocalPath() (string, error) {
...
 }
 
+func cloneURLAndRef(url *url.URL) (*url.URL, string) {
+	localURL := *url
+	ref := localURL.Fragment
+	localURL.Fragment = ""
+	return &localURL, ref
+}
+
+func (r *SourceRepository) DetectAuth() error {
```
Comment (smarterclayton) [inline pkg/generate/app/sourcelookup.go]: Needs some godoc.  Also, this looks pretty fragile, deserves a lot of testing.

**Comment by csrwng** [comment]
comments addressed so far. We can do tests where we check simple credentials and a proxy... however ssh auth is going to be harder. Tests coming shortly.

File: pkg/generate/app/sourcelookup.go (hunk 2 above)
Comment (csrwng) [inline pkg/generate/app/sourcelookup.go]: Fixed

**Comment by csrwng** [comment]
Added integration tests to validate that credential/proxy is reported correctly. @smarterclayton your comments should be addressed.
[test]

**Comment by csrwng** [comment]
waiting on #10615 to merge

**Comment by csrwng** [comment]
[merge]

**Comment by wanghaoran1988** [comment]
@csrwng Would you like to and an extended test and using the private remote git repo for this?

**Pull request merged**